test/e2e/httpproxy/001_required_field_validation_test.go

```diff
@@ -0,0 +1,150 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build e2e
+
+package httpproxy
+
+import (
+	"context"
+
+	contourv1 "github.com/projectcontour/contour/apis/projectcontour/v1"
+	"github.com/projectcontour/contour/test/e2e"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func testRequiredFieldValidation(fx *e2e.Framework) {
+	t := fx.T()
+	ns := "001-required-field-validation"
+
+	fx.CreateNamespace(ns)
+	defer fx.DeleteNamespace(ns)
+
+	// This HTTPProxy is expressed as an Unstructured because the JSON
+	// tags for the relevant field do not include "omitempty", so when
+	// a typed Go struct is serialized to JSON the field *is* included
+	// and therefore does not fail validation.
+	missingConditionHeaderName := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "projectcontour.io/v1",
+			"kind":       "HTTPProxy",
+			"metadata": map[string]interface{}{
+				"name":      "missing-condition-header-name",
+				"namespace": ns,
+			},
+			"spec": map[string]interface{}{
+				"routes": []map[string]interface{}{
+					{
+						"conditions": []map[string]interface{}{
+							{
+								"header": map[string]interface{}{
+									"present": true,
+								},
+							},
+						},
+						"services": []map[string]interface{}{
+							{
+								"name": "foo",
+								"port": 80,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	err := fx.Client.Create(context.TODO(), missingConditionHeaderName)
+	require.Error(t, err)
+	assert.Contains(t, err.Error(), "spec.routes.conditions.header.name: Required value")
+
+	// This HTTPProxy is expressed as an Unstructured because the JSON
```

Comment:
😱

Reply:
Yeah this isn't the greatest, but necessary right now. If we tweak the kubebuilder validations (e.g. apply a minLength of 1) and/or add `omitempty` JSON tags, then this can go away, but I wanted to take a second to think about those changes before making them since they're in the API package.

Reply:
yeah, probably fine to leave as is and file an issue for the api change

Reply:
Good call I'll do that

Reply:
https://github.com/projectcontour/contour/issues/3674